### PR TITLE
feat: appeals add validators when appealed

### LINKED ISSUE #593: Appeals: add validators when appealed
Transaction is in the accepted state and is appealed.
- New validators are added to the validators set. If there were N validators before, N+2 validators should be added.
- N+2 validators process the appeal.
  - Appeal successful: remove old leader, go to pending state, select new leader from 2N+1 validators. Process the transaction
  - Appeal fail: go to accepted state. Don't update the appeal window time.

### PR DESCRIPTION
**Pull request by kstroobants**
Fixes #593

# What

- Implemented issue description
- Refactored consensus flow into state classes and added a transaction_context that contains data that is useful for other states the transaction will go to and it is not in the database. The reason is to perform loops in the consensus with structured code.
- Added class methods to convert dicts to objects. This is needed because an accepted transaction that is appealed needs to be validated by genvm. The genvm accepts only objects (like transaction.consensus_data.leader_receipt.eq_outputs).
- Added tests.

# Why

- To add more value to the user

# Testing done

- Tested the new feature by adding tests to test_base.py
- Simulated it in the simulator

# Checks

- [x] I have tested this code
- [x] I have reviewed my own PR
- [x] I have created an issue for this PR
- [x] I have set a descriptive PR title compliant with [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)

# Reviewing tips

Run test_base.py + play around in the simulator.

### CODE REVIEW

**Review by cristiam86 — CHANGES_REQUESTED** [review]
Good job! there are some issues but, in general, it is very good. Please check with @kirilaa if when the transaction enters the appeal process, its process goes to COMMITTING and REVEALING again or if this is processed in a separated flow

**Comment by kstroobants** [comment]
> Good job! there are some issues but, in general, it is very good. Please check with @kirilaa if when the transaction enters the appeal process, its process goes to COMMITTING and REVEALING again or if this is processed in a separated flow

I checked with him. There is a commit and reveal by the N+2 new validators (that is what the code does).

File: backend/consensus/base.py [hunk 1]
```diff
@@ -562,33 +282,517 @@ async def _appeal_window(self):
...
+        votes = {leader["address"]: leader_receipt.vote.value}
+
+        context.consensus_data.votes = votes
+        context.consensus_data.leader_receipt = leader_receipt
+        context.consensus_data.validators = []
+        context.transactions_processor.set_transaction_result(
+            context.transaction.hash, context.consensus_data.to_dict()
+        )
+
+        ConsensusAlgorithm.dispatch_transaction_status_update(
```
[the remark this answers is not in the document]
Comment (kstroobants) [reply backend/consensus/base.py:532]: I moved the transaction database status to the top of the state classes for proposing and committing to be in line with the sol code.

File: backend/consensus/base.py [hunk 2]
```diff
@@ -562,33 +282,517 @@ async def _appeal_window(self):
...
+    async def handle(self, context):
+        ConsensusAlgorithm.dispatch_transaction_status_update(
+            context.transactions_processor,
+            context.transaction.hash,
+            TransactionStatus.REVEALING,
+            context.msg_handler,
+        )
+
+        for i, validation_result in enumerate(context.validation_results):
+            context.votes[context.validator_nodes[i].address] = (
```
[the remark this answers is not in the document]
Comment (kstroobants) [reply backend/consensus/base.py:595]: context.votes = stores the vote from each validator node (= all votes)
single_reveal_votes = creates a dictionary of votes for the current reveal so the rollup transaction contains leader  vote and 1 validator vote (for each validator).

File: backend/consensus/base.py [hunk 3]
```diff
@@ -562,33 +282,517 @@ async def _appeal_window(self):
...
+                    context.transaction.consensus_data.votes | context.votes
+                )
+                context.consensus_data.validators = (
+                    context.transaction.consensus_data.validators
+                    + context.validation_results
+                )
+                context.transactions_processor.set_transaction_result(
+                    context.transaction.hash, context.consensus_data.to_dict()
+                )
+                ConsensusAlgorithm.dispatch_transaction_status_update(
```
[the remark this answers is not in the document]
Comment (kstroobants) [reply backend/consensus/base.py:639]: @cristiam86 set_transaction_result() on line 636 will write the consensus data to the transaction in the database. In the pending state the consensus mechanism will know all the validators from that stored consensus data. Then it performs the next states where the consensus data will be overwritten with what currently matters in processing the transaction. 

Why do you want to review it? I see two option:
1. Add the transaction consensus data in a history field in the database, then the frontend (or test print) gets the data from the database.
2. Make a memory in the frontend to keep track of the history.

Lets add this to a new issue?

File: tests/unit/consensus/test_helpers.py [hunk 4]
```diff
@@ -165,40 +171,126 @@ def get_nodes_specs(number_of_nodes: int):
...
-            transaction = dict_to_transaction(transaction)
+            transaction = Transaction.from_dict(transaction)
             if not transaction.appealed:
                 if (
                     int(time.time()) - transaction.timestamp_accepted
                 ) > DEFAULT_FINALITY_WINDOW:
-                    consensus.finalize_transaction(
-                        transaction,
-                        transactions_processor,
+                    context = TransactionContext(
```
[the remark this answers is not in the document]
Comment (kstroobants) [reply tests/unit/consensus/test_helpers.py:196]: Me too. It is not future proof. I'll make an issue to investigate this.

**Comment by kstroobants** [comment]
@cristiam86: I changed env finality window to 10 seconds, we can change it back to 30min in the future when the user can easily control it. Okay?

File: backend/consensus/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (cristiam86) [reply backend/consensus/base.py:644]: @kstroobants yes please, let's record all the consensus history in the transaction on another issue: https://github.com/yeagerai/genlayer-studio/issues/715

**Review by cristiam86 — APPROVED** [review]

**Pull request merged**